File: app/src/main/java/com/sorien/ppbthome/IBeaconAdvertiseDataBuilder.java
package com.sorien.ppbthome;

import android.bluetooth.le.AdvertiseData;

import java.nio.ByteBuffer;
import java.util.UUID;

public class IBeaconAdvertiseDataBuilder {

    private final UUID Uuid;
    private final short Major;
    private final short Minor;
    private final byte Tx;

    public IBeaconAdvertiseDataBuilder(UUID Uuid, short Major, short Minor, byte Tx) {
        this.Uuid = Uuid;
        this.Major = Major;
        this.Minor = Minor;
        this.Tx = Tx;
    }

    private static byte[] asBytes(UUID uuid) {
        ByteBuffer bb = ByteBuffer.wrap(new byte[16]);
        bb.putLong(uuid.getMostSignificantBits());
        bb.putLong(uuid.getLeastSignificantBits());
        return bb.array();
    }

    public AdvertiseData build() {
        AdvertiseData.Builder builder = new AdvertiseData.Builder();
        ByteBuffer data = ByteBuffer.allocate(24);
        byte[] uuid = asBytes(Uuid);
        data.put(0, (byte) 0x02); // Beacon Identifier
        data.put(1, (byte) 0x15); // Beacon Identifier
        for (int i = 2; i <= 17; i++) {
            data.put(i, uuid[i - 2]); // adding the UUID
        }
        data.put(18, (byte)(Major >>> 8)); // first byte of Major
        data.put(19, (byte)Major); // second byte of Major
        data.put(20, (byte)(Minor >>> 8)); // first minor
        data.put(21, (byte)Minor); // second minor
        data.put(22, Tx); // txPower
        builder.addManufacturerData(0x4c, data.array()); // using google's company ID
        return builder.build();
    }
}

File: app/src/main/java/com/sorien/ppbthome/IBeaconAdvertiseDataBuilder.kt
package com.sorien.ppbthome

import android.bluetooth.le.AdvertiseData
import com.sorien.ppbthome.IBeaconAdvertiseDataBuilder
import java.nio.ByteBuffer
import java.util.*

class IBeaconAdvertiseDataBuilder(private val Uuid: UUID, private val Major: Short, private val Minor: Short, private val Tx: Byte) {
    fun build(): AdvertiseData {
        val builder = AdvertiseData.Builder()
        val data = ByteBuffer.allocate(24)
        val uuid = asBytes(Uuid)
        data.put(0, 0x02.toByte()) // Beacon Identifier
        data.put(1, 0x15.toByte()) // Beacon Identifier
        for (i in 2..17) {
            data.put(i, uuid[i - 2]) // adding the UUID
        }
        data.put(18, (Major ushr 8).toByte()) // first byte of Major
        data.put(19, Major.toByte()) // second byte of Major
        data.put(20, (Minor ushr 8).toByte()) // first minor
        data.put(21, Minor.toByte()) // second minor
        data.put(22, Tx) // txPower
        builder.addManufacturerData(0x4c, data.array()) // using google's company ID
        return builder.build()
    }

    companion object {
        private fun asBytes(uuid: UUID): ByteArray {
            val bb = ByteBuffer.wrap(ByteArray(16))
            bb.putLong(uuid.mostSignificantBits)
            bb.putLong(uuid.leastSignificantBits)
            return bb.array()
        }
    }
}